### PR TITLE
fix : prevent key collision on KeyValueList add()

### LINKED ISSUE #900: Prevent encoding collision on KeyValueList
In `KeyValueList`:
- if `key` is `_MAX_KEY`, and `value` is zero, `pack` will return `0`
- The encoding here is the same as for the "empty" value, where unpack returns (0,0)
So if the list has the MAX_KEY, 0  element, it will be considered like an empty entry.

Proposed solution : 
`add()` should require `key < _MAX_KEY` (instead of `<=`)

Note : 
This issue should not be present in the current Spoke version, because we are using `uint24` for the key (and max `_MAX_KEY` is `2**32-1`).

### PR DESCRIPTION
**Pull request by Kogaroshi**
- Prevent key collision on `KeyValueList` add() function by requiring `key < _MAX_KEY` (instead of <=)
- This issue anyway should not be present in the current Spoke version, because we are using `uint24` for the key (and max _MAX_KEY is 2**32-1).

Fixes #900

### CODE REVIEW

File: tests/unit/KeyValueList.t.sol [hunk 1]
```diff
@@ -5,9 +5,27 @@ pragma solidity ^0.8.0;
 import {Test} from 'forge-std/Test.sol';
 import {KeyValueList} from 'src/spoke/libraries/KeyValueList.sol';
 
+import {console2 as console} from 'forge-std/console2.sol';
```
Comment (DhairyaSethi) [inline tests/unit/KeyValueList.t.sol]: pls rm

File: tests/unit/KeyValueList.t.sol [hunk 2]
```diff
@@ -119,3 +137,18 @@ contract KeyValueListTest is Test {
     return value % KeyValueList._MAX_VALUE;
   }
 }
+
+/// @dev needed to catch the emitted error from the library
```
Comment (DhairyaSethi) [inline tests/unit/KeyValueList.t.sol]: can add setting (we also use it for math lib tests) to avoid needing this
```
/// forge-config: default.allow_internal_expect_revert = true
```

File: tests/unit/KeyValueList.t.sol (hunk 1 above)
Comment (Kogaroshi) [reply tests/unit/KeyValueList.t.sol]: oops mb

File: src/spoke/libraries/KeyValueList.sol [hunk 3]
```diff
@@ -36,7 +36,7 @@ library KeyValueList {
 
   /// @notice Inserts packed `key`, `value` at `idx`. Reverts if data exceeds maximum allowed size.
   function add(List memory self, uint256 idx, uint256 key, uint256 value) internal pure {
-    require(key <= _MAX_KEY && value <= _MAX_VALUE, MaxDataSizeExceeded());
+    require(key < _MAX_KEY && value <= _MAX_VALUE, MaxDataSizeExceeded());
```
Comment (DhairyaSethi) [inline src/spoke/libraries/KeyValueList.sol]: let's limit value to also be strictly less than MAX_VALUE for consistency? doesn't really matter, maybe we rename constants?

File: tests/unit/KeyValueList.t.sol [hunk 4]
```diff
@@ -5,9 +5,25 @@ pragma solidity ^0.8.0;
 import {Test} from 'forge-std/Test.sol';
 import {KeyValueList} from 'src/spoke/libraries/KeyValueList.sol';
 
+/// forge-config: default.allow_internal_expect_revert = true
 contract KeyValueListTest is Test {
   using KeyValueList for KeyValueList.List;
 
+  function test_fuzz_add(uint256 key, uint256 value) public {
```
Comment (DhairyaSethi) [inline tests/unit/KeyValueList.t.sol:87]: let's write a test where we `add` unique values only and assert only unique values are received on `get`?

File: src/spoke/libraries/KeyValueList.sol (hunk 3 above)
Comment (Kogaroshi) [reply src/spoke/libraries/KeyValueList.sol]: I'm okay with current constant names

File: src/spoke/libraries/KeyValueList.sol [hunk 5]
```diff
@@ -36,7 +36,7 @@ library KeyValueList {
 
   /// @notice Inserts packed `key`, `value` at `idx`. Reverts if data exceeds maximum allowed size.
   function add(List memory self, uint256 idx, uint256 key, uint256 value) internal pure {
-    require(key <= _MAX_KEY && value <= _MAX_VALUE, MaxDataSizeExceeded());
+    require(key < _MAX_KEY && value < _MAX_VALUE, MaxDataSizeExceeded());
```
Comment (miguelmtzinf) [inline src/spoke/libraries/KeyValueList.sol:40]: should this revert in `pack` instead? also, let's add a dev docs above

File: tests/unit/KeyValueList.t.sol [hunk 6]
```diff
@@ -5,9 +5,79 @@ pragma solidity ^0.8.0;
...
+  function test_add_unique() public {
+    KeyValueList.List memory list = KeyValueList.init(9);
+
+    list.add(0, 1, 1);
+    list.add(1, 100, 1e15);
+    list.add(2, 100, 5e20);
+    list.add(3, 5e4, 5e20);
+    list.add(4, 5e4, 1e12);
+    list.add(5, 45e6, 2.5e50);
+    list.add(6, 45e6, 10);
```
Comment (DhairyaSethi) [inline tests/unit/KeyValueList.t.sol]: pls add edge cases in the explicit case, ie type(uint32).max - 1; type(uint224).max - 1 etc

File: tests/unit/KeyValueList.t.sol [hunk 7]
```diff
@@ -5,9 +5,79 @@ pragma solidity ^0.8.0;
 import {Test} from 'forge-std/Test.sol';
 import {KeyValueList} from 'src/spoke/libraries/KeyValueList.sol';
 
+/// forge-config: default.allow_internal_expect_revert = true
 contract KeyValueListTest is Test {
   using KeyValueList for KeyValueList.List;
 
+  function test_add_unique() public {
```
Comment (DhairyaSethi) [inline tests/unit/KeyValueList.t.sol:12]: let's have a fuzz test for this as well?
fuzzed input can just be seed or array whichever is better wrt performance

File: src/spoke/libraries/KeyValueList.sol [hunk 8]
```diff
@@ -35,8 +35,9 @@ library KeyValueList {
   }
 
   /// @notice Inserts packed `key`, `value` at `idx`. Reverts if data exceeds maximum allowed size.
```
Comment (yan-man) [inline src/spoke/libraries/KeyValueList.sol:37]: should we update the natspec here or move below into @dev too?

File: tests/unit/KeyValueList.t.sol [hunk 9]
```diff
@@ -5,9 +5,106 @@ pragma solidity ^0.8.0;
...
+    list.add(3, 5e4, 5e20);
+    list.add(4, 5e4, 1e12);
+    list.add(5, 45e6, 2.5e50);
+    list.add(6, 45e6, 10);
+    vm.expectRevert(KeyValueList.MaxDataSizeExceeded.selector);
+    list.add(7, 4294967295, 10000000000);
+    vm.expectRevert(KeyValueList.MaxDataSizeExceeded.selector);
+    list.add(7, 45e8, 10000000000);
+    vm.expectRevert(KeyValueList.MaxDataSizeExceeded.selector);
+    list.add(type(uint32).max - 1, 75e9, 10000000000);
```
Comment (DhairyaSethi) [inline tests/unit/KeyValueList.t.sol]: uint32 max for index? 

i thought the suggestion was uint32.max at KEY would revert, and we test the edge case when key is uint32.max - 1 which doesn't revert

Comment (Kogaroshi) [reply tests/unit/KeyValueList.t.sol]: indeed, shouldn't fail is key is `uint32.max - 1`, here it fails because I put it as the index ...
